### PR TITLE
Describe partial decode support for codecs

### PR DESCRIPTION
**Pull request by jbms**
This is a draft of the spec change to resolve issue https://github.com/zarr-developers/zarr-specs/issues/220.

### CODE REVIEW

**Review by normanrz — COMMENTED** [review]
Looks pretty good. I would note for each registered codec whether it is {bytes,array} -> {bytes,array}. Also, we could add to blosc and sharding codecs that partial decoding is recommended for implementations.

**Comment by jstriebel** [comment]
I agree with @normanrz, every codex should declare what its inputs and outputs are. Apart from that LGTM 👍

**Comment by jbms** [comment]
I rebased this PR, and added a description for each codec of what type of codec it is.

File: docs/v3/codecs/transpose/v1.0.rst [hunk 1]
```diff
@@ -71,9 +72,7 @@ order:
 Format and algorithm
 ====================
 
-The decoded chunk representation to which this codec is applied must be an
-array.  Implementations must fail if this codec is specified immediately after
-another codec that produces a byte string as its encoded representation.
+This is an ``array -> array` codec.
```
Comment (normanrz) [inline docs/v3/codecs/transpose/v1.0.rst:75]:
```suggestion
This is an ``array -> array`` codec.
```

File: docs/v3/core/v3.0.rst [hunk 2]
```diff
@@ -1089,6 +1089,36 @@ each of these two representations are defined to be either:
...
+the following form:
+
+- zero or more ``array -> array`` codecs; followed by
+- at most one ``array -> bytes`` codec; followed by
+- zero or more ``bytes -> bytes`` codecs.
+
+If no ``array -> bytes`` codec is specified, then the default byte
+representation for the data type of the array is used.  For all data types
+currently defined by the core spec, that is equivalent to the ``endian`` codec
+with an endianness of ``little``.
```
Comment (normanrz) [inline docs/v3/core/v3.0.rst:1120]: Should we also mention the `transpose` codec with order `C`?

Comment (jbms) [reply docs/v3/core/v3.0.rst:1120]: The endian codec with endianness of little may be a noop in terms of the actual in-memory representation of the data, but it does have an effect in terms of the logical data model, in that it transforms a multidimensional array into a (one-dimensional) byte sequence.  In contrast, the transpose codec with order [0, 1, ..., rank-1], i.e. "C", is logically an identity transform.  It is the endian codec that uses C order, and if you want a different order you have to transpose it first.

**Review by normanrz — APPROVED** [review]

**Review by jstriebel — APPROVED** [review]
LGTM 👍

**Pull request merged**